File: 04b.js
import * as fs from 'fs';

function calculate_score(board, number)
{
    var score = 0;
    board.forEach(line => {
        line.forEach(item => {
            if (item !== '-1') {
                score += parseInt(item);
            }
        });
    });
    return score*parseInt(number);
}

function mark_number(boards, number)
{
    boards.forEach(board => {
        board.forEach(line => {
            line.forEach((item, index) => {
                if (item == number) {
                    line[index] = '-1';
                }
            });
        });
    });
}

function check_for_bingo(boards, number, wins)
{
    var already_won = wins.map(item => item[0]);
    boards.forEach((board, board_index) => {
        if (!(already_won.includes(board_index))) {
            board.forEach(line => {
                var marked = 0
                line.forEach(item => {
                    if (item === '-1') {
                        marked++;
                    }
                });
                if (marked === 5) { // bingo
                    var score = calculate_score(board, number);
                    wins.push([board_index,score]);
                }
            });
            board[0].forEach((item, index) => {
                var marked = 0;
                board.forEach(line => {
                    if (line[index] === '-1') {
                        marked++;
                    }
                });
                if (marked === 5) { // bingo
                    var score = calculate_score(board, number);
                    wins.push([board_index,score]);
                }
            });    
        }
    });
}

const filename = "04.dat";
fs.readFile(filename, (err, data) => {
    if (err) throw err;

    var board_data = data.toString().split('\n\n');
    var lottery = board_data[0];
    board_data = board_data.slice(1);
    
    var boards = [];
    board_data.forEach(board_section => {
        var board = [];
        board_section.split('\n').forEach(line => {
            var match = [...line.matchAll(/\d+/g)];
            if (match) {
                var board_line = match;
                board.push(board_line);
            }
        });
        boards.push(board);
    });
    
    var wins = [];
    lottery.split(',').forEach(number => {
        mark_number(boards, number);
        check_for_bingo(boards, number, wins);
    });
    console.log(wins[wins.length - 1][1])
});

File: 04b.py
#!/usr/bin/python3

import re

def calculate_score(board: list[list[str]], number: str) -> int:
    score = 0
    for line in board:
        for item in line:
            if item != '-1':
                score += int(item)
    return score*int(number)

def mark_number(boards: list[list[list[str]]], number: str) -> None:
    for board in boards:
        for line in board:
            for index, item in enumerate(line):
                if item == number:
                    line[index] = '-1'

def check_for_bingo(boards: list[list[list[str]]], number: str, wins: list[(int,int)]) -> None:
    already_won = [item[0] for item in wins]
    for board_index, board in enumerate(boards):
        if board_index in already_won:
            continue
        for line in board:
            marked = 0
            for item in line:
                if item == '-1':
                    marked += 1
            if marked == 5: # bingo
                score = calculate_score(board, number)
                wins.append((board_index,score))
        for index, _ in enumerate(board[0]):
            marked = 0
            for line in board:
                if line[index] == '-1':
                    marked += 1
            if marked == 5: # bingo
                score = calculate_score(board, number)
                wins.append((board_index, score))

with open('04.dat') as f:
    data = f.read()
board_data = data.split('\n\n')
lottery = board_data[0]
board_data = board_data[1:]

boards = []
for board_section in board_data:
    board = []
    for line in board_section.split('\n'):
        match = re.findall(r'\d+', line)
        if (match):
            board_line = match
            board.append(board_line)
    boards.append(board)

wins = []
for number in lottery.split(','):
    mark_number(boards, number)
    check_for_bingo(boards, number, wins)
print(wins[len(wins) - 1][1])